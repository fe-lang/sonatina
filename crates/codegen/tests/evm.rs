use dir_test::{dir_test, Fixture};

use hex::ToHex;
use revm::{
    inspector_handle_register,
    interpreter::Interpreter,
    primitives::{
        AccountInfo, Address, Bytecode, Bytes, CancunSpec, Env, ExecutionResult, TransactTo, U256,
    },
    Context, EvmContext, Handler,
};

use sonatina_codegen::{
    critical_edge::CriticalEdgeSplitter,
    domtree::DomTree,
    isa::evm::{opcode::OpCode, EvmBackend},
    liveness::Liveness,
    machinst::{assemble::ObjectLayout, lower::Lower, vcode::VCode},
    stackalloc::SimpleAlloc,
};
use sonatina_ir::{
    cfg::ControlFlowGraph,
    ir_writer::{FuncWriteCtx, IrWrite},
    isa::evm::Evm,
    module::ModuleCtx,
    BlockId, Function,
};
use sonatina_parser::{parse_module, ParsedModule};
use sonatina_triple::{Architecture, OperatingSystem, Vendor};
use std::io::{stderr, Write};

// XXX copied from fe test-utils
#[macro_export]
macro_rules! snap_test {
    ($value:expr, $fixture_path: expr) => {
        let mut settings = insta::Settings::new();
        let fixture_path = ::std::path::Path::new($fixture_path);
        let fixture_dir = fixture_path.parent().unwrap();
        let fixture_name = fixture_path.file_stem().unwrap().to_str().unwrap();

        settings.set_snapshot_path(fixture_dir);
        settings.set_input_file($fixture_path);
        settings.set_prepend_module_to_snapshot(false);
        settings.bind(|| {
            insta::_macro_support::assert_snapshot(
                (insta::_macro_support::AutoName, $value.as_str()).into(),
                std::path::Path::new(env!("CARGO_MANIFEST_DIR")),
                fixture_name,
                module_path!(),
                file!(),
                line!(),
                stringify!($value),
            )
            .unwrap()
        })
    };
}

fn parse_sona(content: &str) -> ParsedModule {
    match parse_module(content) {
        Ok(module) => module,
        Err(errs) => {
            let mut w = stderr();
            for err in errs {
                err.print(&mut w, "[test]", content, true).unwrap();
            }
            panic!("Failed to parse test file. See errors above.")
        }
    }
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/evm",
    glob: "*.sntn"
)]
fn test_evm(fixture: Fixture<&str>) {
    let parsed = parse_sona(fixture.content());

    let backend = EvmBackend::new(Evm::new(sonatina_triple::TargetTriple {
        architecture: Architecture::Evm,
        vendor: Vendor::Ethereum,
        operating_system: OperatingSystem::Evm(sonatina_triple::EvmVersion::Cancun),
    }));

    let mut v = Vec::new();
    let funcs = parsed
        .debug
        .func_order
        .iter()
        .map(|fref| {
            let (vcode, block_order) = parsed.module.func_store.modify(*fref, |function| {
                let (vcode, block_order) = vcode_for_fn(function, &parsed.module.ctx, &backend);
                let ctx = FuncWriteCtx::with_debug_provider(function, *fref, &parsed.debug);
                vcode.write(&mut v, &ctx).unwrap();
                (vcode, block_order)
            });
            writeln!(v).unwrap();
            (*fref, vcode, block_order)
        })
        .collect();

    write!(&mut v, "\n\n---------------\n\n").unwrap();

    let mut layout = ObjectLayout::new(funcs, 0);
    let mut i = 0;
    while layout.resize(&backend, 0) {
        i += 1;
        println!("resize iteration {i}");
    }
    layout.print(&mut v, &parsed.module, &parsed.debug).unwrap();

    let mut bytecode = Vec::new();
    layout.emit(&backend, &mut bytecode);
    let hex = bytecode.encode_hex::<String>();

    writeln!(&mut v, "\n0x{hex}").unwrap();

    let (res, trace) = run_on_evm(&bytecode);

    writeln!(&mut v, "\n{trace}").unwrap();
    writeln!(&mut v, "\n{res:?}").unwrap();

    snap_test!(String::from_utf8(v).unwrap(), fixture.path());
}

fn run_on_evm(bytecode: &[u8]) -> (ExecutionResult, String) {
    let mut db = revm::InMemoryDB::default();
    let revm_bytecode = Bytecode::new_raw(Bytes::copy_from_slice(bytecode));
    let test_address = Address::repeat_byte(0x12);
    db.insert_account_info(
        test_address,
        AccountInfo {
            balance: U256::ZERO,
            nonce: 0,
            code_hash: revm_bytecode.hash_slow(),
            code: Some(revm_bytecode),
        },
    );

    let mut env = Env::default();
    env.tx.clear();
    env.tx.transact_to = TransactTo::Call(test_address);
    env.tx.data = Bytes::copy_from_slice(&[0, 0, 0, 11, 0, 0, 0, 22]);

    let context = Context::new(
        EvmContext::new_with_env(db, Box::new(env)),
        // TracingInspector::new(TracingInspectorConfig::all());
        TestInspector::new(vec![]),
    );

    let mut evm = revm::Evm::new(context, Handler::mainnet::<CancunSpec>());
    evm = evm
        .modify()
        .append_handler_register(inspector_handle_register)
        .build();

    let res = evm.transact_commit();

    // let mut w = TraceWriter::new(&mut v).use_colors(ColorChoice::Never);
    // w.write_arena(evm.context.external.traces()).unwrap();

    match res.clone() {
        Ok(r) => (r, String::from_utf8(evm.context.external.w).unwrap()),
        Err(e) => panic!("evm failure: {e}"),
    }
}

struct TestInspector<W: Write> {
    w: W,
}

impl<W: Write> TestInspector<W> {
    fn new(w: W) -> TestInspector<W> {
        Self { w }
    }
}

impl<W: Write, DB: revm::Database> revm::Inspector<DB> for TestInspector<W> {
    fn initialize_interp(&mut self, _interp: &mut Interpreter, _context: &mut EvmContext<DB>) {
        writeln!(
            self.w,
            "{:>6}  {:<17} input (stack grows to the right)",
            "pc", "opcode"
        )
        .unwrap();
    }

    fn step(&mut self, interp: &mut Interpreter, _context: &mut EvmContext<DB>) {
        // xxx tentatively writing input stack; clean up
        let pc = interp.program_counter();

        let op = interp.current_opcode();
        let code = unsafe { std::mem::transmute::<u8, revm::interpreter::OpCode>(op) };

        let stack = interp.stack().data();

        write!(
            self.w,
            "{:>6}  {:0>2}  {:<12}  ",
            pc,
            format!("{op:x}"),
            code.info().name(),
        )
        .unwrap();
        let imm_size = code.info().immediate_size() as usize;
        if imm_size > 0 {
            let imm_bytes = interp.bytecode.slice((pc + 1)..(pc + 1 + imm_size));
            writeln!(self.w, "{}", imm_bytes).unwrap();
        } else {
            writeln!(self.w, "{stack:?}").unwrap();
        }
    }

    fn step_end(&mut self, _interp: &mut Interpreter, _context: &mut EvmContext<DB>) {
        // NOTE: annoying revm behavior: `interp.current_opcode()` now returns the next opcode.
    }
}

fn vcode_for_fn(
    function: &mut Function,
    module: &ModuleCtx,
    backend: &EvmBackend,
) -> (VCode<OpCode>, Vec<BlockId>) {
    let mut cfg = ControlFlowGraph::new();
    cfg.compute(function);
    let mut splitter = CriticalEdgeSplitter::new();
    splitter.run(function, &mut cfg);

    let mut liveness = Liveness::new();
    liveness.compute(function, &cfg);
    let mut dom = DomTree::new();
    dom.compute(&cfg);
    let mut alloc = SimpleAlloc::for_function(function, &cfg, &dom, &liveness, 16);
    let lower = Lower::new(module, function);

    match lower.lower(backend, &mut alloc) {
        Ok(vcode) => (vcode, dom.rpo().to_owned()),
        Err(err) => panic!("{:?}", err),
    }
}
