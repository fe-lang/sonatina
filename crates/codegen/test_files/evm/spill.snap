---
source: crates/codegen/tests/evm.rs
expression: "String::from_utf8(v).unwrap()"
input_file: test_files/evm/spill.sntn
---
// func public %sum8(v0.i64, v1.i64, v2.i64, v3.i64, v4.i64, v5.i64, v6.i64, v7.i64) -> i64
sum8:
  block0:
    JUMPDEST
    ADD  // v8.i64 = add v0 v1;
    SWAP2  // v9.i64 = add v2 v3;
    SWAP1
    ADD
    SWAP3  // v10.i64 = add v4 v5;
    DUP3
    ADD
    SWAP5  // v11.i64 = add v6 v7;
    DUP5
    ADD
    SWAP3  // v12.i64 = add v8 v9;
    SWAP1
    ADD
    SWAP1  // v13.i64 = add v10 v11;
    POP
    SWAP1
    DUP4
    ADD
    SWAP1  // v14.i64 = add v12 v13;
    ADD
    SWAP2  // return v14;
    POP
    POP
    SWAP1
    JUMP

// func public %spill(v0.i64, v1.i64, v2.i64, v3.i64, v4.i64, v5.i64, v6.i64, v7.i64) -> i64
spill:
  block1:
    DUP8
    PUSH0
    MLOAD
    MSTORE
    JUMPDEST
    DUP2  // v8.i64 = add v0 v1;
    DUP2
    ADD
    DUP5  // v9.i64 = add v2 v3;
    DUP5
    ADD
    DUP8  // v10.i64 = add v4 v5;
    DUP8
    ADD
    DUP11  // v11.i64 = add v6 v7;
    DUP11
    ADD
    DUP3  // v12.i64 = add v8 v9;
    DUP5
    ADD
    DUP2  // v13.i64 = add v10 v11;
    DUP4
    ADD
    DUP1  // v14.i64 = add v12 v13;
    DUP3
    ADD
    DUP1  // jump block2;
    PUSH1 block2
    JUMP
  block2:
    JUMPDEST
    PUSH1 `pc + (11)`  // v16.i64 = call %sum8 v0 v1 v2 v3 v4 v5 v6 v7;
    DUP16
    DUP16
    DUP16
    DUP16
    DUP16
    DUP16
    DUP16
    DUP16
    PUSH1 FuncRef(0)
    JUMP
    JUMPDEST
    PUSH1 `pc + (11)`  // v17.i64 = call %sum8 v8 v9 v10 v11 v12 v13 v14 v15;
    SWAP1
    DUP3
    DUP5
    DUP7
    DUP9
    DUP11
    DUP13
    DUP15
    PUSH1 FuncRef(0)
    JUMP
    JUMPDEST
    SWAP1  // v18.i64 = add v16 v17;
    ADD
    DUP1  // v19.i64 = add v7 v18;
    PUSH0
    MLOAD
    MLOAD
    ADD
    PUSH2 0x3e8 (1000)  // v20.i1 = gt v19 1000.i64;
    DUP2
    GT
    PUSH1 block3  // br v20 block3 block4;
    JUMPI
    PUSH1 block4
    JUMP
  block3:
    JUMPDEST
    POP  // return v18;
    SWAP15
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    POP
    SWAP1
    JUMP
  block4:
    JUMPDEST
    SWAP1  // jump block2;
    POP
    PUSH1 block2
    JUMP



---------------

// func public %sum8(v0.i64, v1.i64, v2.i64, v3.i64, v4.i64, v5.i64, v6.i64, v7.i64) -> i64
sum8:
  block0:
    0    JUMPDEST
    1    ADD  // v8.i64 = add v0 v1;
    2    SWAP2  // v9.i64 = add v2 v3;
    3    SWAP1
    4    ADD
    5    SWAP3  // v10.i64 = add v4 v5;
    6    DUP3
    7    ADD
    8    SWAP5  // v11.i64 = add v6 v7;
    9    DUP5
   10    ADD
   11    SWAP3  // v12.i64 = add v8 v9;
   12    SWAP1
   13    ADD
   14    SWAP1  // v13.i64 = add v10 v11;
   15    POP
   16    SWAP1
   17    DUP4
   18    ADD
   19    SWAP1  // v14.i64 = add v12 v13;
   20    ADD
   21    SWAP2  // return v14;
   22    POP
   23    POP
   24    SWAP1
   25    JUMP
// func public %spill(v0.i64, v1.i64, v2.i64, v3.i64, v4.i64, v5.i64, v6.i64, v7.i64) -> i64
spill:
  block1:
   26    DUP8
   27    PUSH0
   28    MLOAD
   29    MSTORE
   30    JUMPDEST
   31    DUP2  // v8.i64 = add v0 v1;
   32    DUP2
   33    ADD
   34    DUP5  // v9.i64 = add v2 v3;
   35    DUP5
   36    ADD
   37    DUP8  // v10.i64 = add v4 v5;
   38    DUP8
   39    ADD
   40    DUP11  // v11.i64 = add v6 v7;
   41    DUP11
   42    ADD
   43    DUP3  // v12.i64 = add v8 v9;
   44    DUP5
   45    ADD
   46    DUP2  // v13.i64 = add v10 v11;
   47    DUP4
   48    ADD
   49    DUP1  // v14.i64 = add v12 v13;
   50    DUP3
   51    ADD
   52    DUP1  // jump block2;
   53    PUSH1 56 (block2)
   55    JUMP
  block2:
   56    JUMPDEST
   57    PUSH1 69  // v16.i64 = call %sum8 v0 v1 v2 v3 v4 v5 v6 v7;
   59    DUP16
   60    DUP16
   61    DUP16
   62    DUP16
   63    DUP16
   64    DUP16
   65    DUP16
   66    DUP16
   67    PUSH0 0 (FuncRef(0))
   68    JUMP
   69    JUMPDEST
   70    PUSH1 82  // v17.i64 = call %sum8 v8 v9 v10 v11 v12 v13 v14 v15;
   72    SWAP1
   73    DUP3
   74    DUP5
   75    DUP7
   76    DUP9
   77    DUP11
   78    DUP13
   79    DUP15
   80    PUSH0 0 (FuncRef(0))
   81    JUMP
   82    JUMPDEST
   83    SWAP1  // v18.i64 = add v16 v17;
   84    ADD
   85    DUP1  // v19.i64 = add v7 v18;
   86    PUSH0
   87    MLOAD
   88    MLOAD
   89    ADD
   90    PUSH2 0x3e8 (1000)  // v20.i1 = gt v19 1000.i64;
   93    DUP2
   94    GT
   95    PUSH1 101 (block3)  // br v20 block3 block4;
   97    JUMPI
   98    PUSH1 121 (block4)
  100    JUMP
  block3:
  101    JUMPDEST
  102    POP  // return v18;
  103    SWAP15
  104    POP
  105    POP
  106    POP
  107    POP
  108    POP
  109    POP
  110    POP
  111    POP
  112    POP
  113    POP
  114    POP
  115    POP
  116    POP
  117    POP
  118    POP
  119    SWAP1
  120    JUMP
  block4:
  121    JUMPDEST
  122    SWAP1  // jump block2;
  123    POP
  124    PUSH1 56 (block2)
  126    JUMP


0x5b01919001928201948401929001905090830190019150509056875f51525b8181018484018787018a8a01828401818301808201806038565b60458f8f8f8f8f8f8f8f5f565b605290828486888a8c8e5f565b9001805f5151016103e881116065576079565b509e50505050505050505050505050505090565b9050603856
